src/vision/status_quo/barbara_builds_an_async_executor.md

```diff
@@ -0,0 +1,168 @@
+
+Barbara wants to set priorities to the tasks spawned to the executor. However, she finds no existing async executor provides such a feature so she decided to build her own async executor.
+
+First, Barbara found [crossbeam-deque](https://crates.io/crates/crossbeam-deque) provides work-stealing deques of good quality. She could use it to build task schedulers. It is not the main part of the story. Each working thread can get a task from the scheduler and poll it in loop.
+
+But wait, what is a "task"?
+
+At first, Barbara thought it must contain the `Future` itself and the additional priority which was used by the scheduler. So she first wrote:
+
+```rust
+pub struct Task {
+    future: Pin<Box<dyn Future<Output = ()> + Send + 'static>>,
+    priority: u8
+}
+```
+
+And the working thread loop should run something like:
+
+```rust
+pub fn poll_task(task: Task) {
+    let waker = todo!();
+    let mut cx = Context::from_waker(&waker);
+    task.future.as_mut().poll(&mut cx);
+}
+```
+
+"How to create a waker?" Barbara asked herself. Quickly, she found the `Wake` trait. Seeing the `wake` method takes an `Arc<Self>`, she realized the task in the scheduler should be stored in an `Arc`. After some thought, she thinks it makes sense because both the deque in the scheduler and the waker may hold a reference to the task.
+
+To implement `Wake`, the `Task` should contain the sender of the scheduler. She changed the code to something like this:
+
+```rust
+pub struct Task {
+    future: Pin<Box<dyn Future<Output = ()> + Send + 'static>>,
+    scheduler: SchedulerSender,
+    priority: u8,
+}
+
+unsafe impl Sync for Task {}
+
+impl Wake for Task {
+    fn wake(self: Arc<Self>) {
+        self.scheduler.send(self.clone());
+    }
+}
+
+pub fn poll_task(task: Arc<Task>) {
+    let waker = Waker::from(task.clone());
+    let mut cx = Context::from_waker(&waker);
+    task.future.as_mut().poll(&mut cx);
+//  ^^^^^^^^^^^ cannot borrow as mutable
+}
+```
+
+The code still needed some change because the `future` in the  `Arc<Task>` became immutable. 
+
+"Okay. I can guarantee `Task` is created from a `Pin<Box<Future>>` and I think the same future will not be polled concurrently in two threads. So let me bypass the safety checks." Barbara confidently used some `unsafe` blocks to make it compile.
+
+```rust
+pub struct Task {
+    future: *mut (dyn Future<Output = ()> + Send + 'static),
+    ...
+}
+
+unsafe impl Send for Task {}
+unsafe impl Sync for Task {}
+
+pub fn poll_task(task: Arc<Task>) {
+    ...
+    unsafe {
+        Pin::new_unchecked(&mut *task.future).poll(&mut cx);
+    }
+}
+```
+
+Luckily, a colleague of Barbara noticed something wrong. The `wake` method could be called multiple times so multiple copies of the task could exist in the scheduler. The scheduler might not work correctly because of it and a more severe problem was that multiple threads might get copies of the same task from the scheduler and cause a race in polling the future.
```

Comment:
```suggestion
Luckily, a colleague of Barbara noticed something wrong. The `wake` method could be called multiple times so multiple copies of the task could exist in the scheduler. The scheduler might not work correctly because of this. What's worse, a more severe problem was that multiple threads might get copies of the same task from the scheduler and cause a race in polling the future.
```